hass-dash.el

```diff
@@ -143,193 +91,136 @@
+This just uses `widget-default-create', but sets the `:tag' property if it isn't
+already set using the `:title' and `:title-face' properties."
+  (unless (widget-get widget :tag)
+    (widget-put widget :tag (propertize (widget-get widget :title)
+                                        'face (widget-get widget :title-face))))
+  (widget-default-create widget))
+
+(defun hass-dash--toggle-widget-value-get (widget)
+  "Get the state for a toggle WIDGET."
+  (string= (hass-state-of (widget-get widget :entity-id)) "on"))
+
+(defun hass-dash--widget-action (widget &optional _)
+  "Action handler for WIDGET.
+If the `:service' property is set, this will call that service.  Otherwise, it
+will call the relevant service in `hass-dash--default-services'.
+
+The `:confirm' property can be used to trigger a confirmation before calling the
+service.  It can take on the following values:
+
+• if it's a string, that string will be used for the prompt
+• if it's a function, then the service will only be called if that function
+  returns t
+• if it's t, a prompt will be created using the value returned by
+  `hass-dash--widget-label'"
+  (let* ((confirm (widget-get widget :confirm))
+         (entity-id (widget-get widget :entity-id))
+         (service (or (widget-get widget :service)
+                      (cdr (assoc (hass--domain-of-entity entity-id)
+                                  hass-dash--default-services)))))
+    (cond ((stringp confirm) (when (y-or-n-p confirm)
+                               (hass-call-service entity-id service nil)))
+          ((functionp confirm) (when (funcall confirm entity-id)
+                                 (hass-call-service entity-id service nil)))
+          (confirm (when (y-or-n-p (concat "Toggle "
+                                           (hass-dash--widget-label widget)
+                                           "? "))
+                     (hass-call-service entity-id service nil)))
+          (t (hass-call-service entity-id service nil)))))
+
+(define-widget 'hass-dash-toggle 'toggle
+  "A toggle widget for home-assistant dashboards.
+You must pass an `:entity-id' property to indicate the id of the entity in Home
+Assistant.  The following optional properties can also be used:
+
+• `:service': The service to call when triggering the action on the widget.  If
+  not passed, then the default will be found in `hass-dash--default-services'
+  instead.
+• `:label': The friendly name to show for the widget.  If not passed, a sane
+  default will be found in the list of available entities.  If nothing is found
+  there, then the `:entity-id' property value will be used.
+• `:icon': The icon to show for the widget.  If not passed one will be found
+  based on the entity id.
+• `:confirm': If passed, this will control how the action is confirmed before
+  being confirmed.  See `hass-dash--widget-action' for details."
+  :create 'hass-dash--widget-create
+  :format "%t: %[[%v]%]\n"
```

Comment:
So this is actually broken because we're rebuilding the whole buffer whenever any update comes in for the tracked entities.  But I think I can get away from having to rebuild the buffer now since I can track an alist that tells me which entity ids map to which widgets on the page.  Will do that and update the PR.